### ISSUE TITLE
Hot-reloading Bevy systems that use Bevy Resources do not work

### ISSUE THREAD

**Issue by kaphula** [rksm/hot-lib-reloader-rs#39]
While hot-reloading simple systems work, when you add Bevy resources the hot-reloading breaks. [Here is a repo/branch to demonstrate this issue](https://github.com/kaphula/hot-lib-reloader-nested-crates/tree/bevy_resource_crash_test), see the `main.rs` file. 

It was suggested to me that since Bevy is a dependency on the crate that will be hot-reloaded, there is another dependency of Bevy, crossbeam, which will force recompilation of all crates of the project due to its buildscripts even though you would only want to recompile the hot-reload crate that contains the magnled systems. This is apparantly messes up the `TypeId`s and the program crashes or ends up with undefined behavior.

**Comment by jakkos-net** [comment]
I don't think this is an issue with `Resource`s, I think it's a problem with `World`. 

I just tried to hot reload a function that looks like:
```
pub fn test_system(mut my_res: ResMut<MyResource>) {
  // do something with the resource
}
```

and it works for me on Bevy 0.15. However if I do:
```
pub fn test_system(world: &mut World) {
  let mut my_res = world.resource_mut::<MyResource>();
  // do something with the resource
}
```
Then it crashes because `MyResource` apparently doesn't exist (when I know it does).

A similar thing happens with `world.query::<&MyComponent>()` where it can't find any entities, but using `Query<&MyComponent>` as a function argument works fine.

As you say, it's probably due to messed up `TypeId`s. I'm a little bit out of my depth here, but if we could find some way to access the "proper" `TypeId`s, we might be able to create something that works like `SystemState` but does all the `TypeId` remapping for you. This would be super handy because it would basically be as powerful as being able to change the system signature during hot reloading 😁
